js/ui/paygUnlockDialog.js

```diff
@@ -0,0 +1,395 @@
+        this._spinner.actor.opacity = 0;
+        this._spinner.actor.show();
+        buttonSpacer.add_child(this._spinner.actor);
+
+        this._nextButton = new St.Button({ style_class: 'modal-dialog-button button',
+                                           button_mask: St.ButtonMask.ONE | St.ButtonMask.THREE,
+                                           reactive: true,
+                                           can_focus: true,
+                                           label: _("Unlock"),
+                                           x_align: St.Align.END,
+                                           y_align: St.Align.END });
+        this._nextButton.add_style_pseudo_class('default');
+        buttonsBox.add_child(this._nextButton);
+
+        return buttonsBox;
+    },
+
+    _onCancelled: function() {
+        this._cancelled = true;
+        this._reset();
+
+        // The ScreenShield will connect to the 'failed' signal
+        // to know when to cancel the unlock dialog.
+        if (this._verificationStatus != UnlockStatus.SUCCEEDED)
+            this.emit('failed');
+    },
+
+    _validateCurrentCode: function() {
+        // The PaygUnlockCodeEntry widget will only accept valid
+        // characters, so we only need to check the length here.
+        return this._entry.length == CODE_ENTRY_REQUIRED_LENGTH;
+    },
+
+    _updateNextButtonSensitivity: function() {
+        let sensitive = this._validateCurrentCode() && this._verificationStatus != UnlockStatus.VERIFYING;
+        this._nextButton.reactive = sensitive;
+        this._nextButton.can_focus = sensitive;
+    },
+
+    _updateSensitivity: function() {
+        this._updateNextButtonSensitivity();
+        this._entry.setEnabled(this._verificationStatus != UnlockStatus.VERIFYING);
+    },
+
+    _setErrorMessage: function(message) {
+        if (message) {
+            this._errorMessage.text = message;
+            this._errorMessage.opacity = 255;
+        } else {
+            this._errorMessage.text = '';
+            this._errorMessage.opacity = 0;
+        }
+    },
+
+    _startSpinning: function() {
+        this._spinner.play();
+        this._spinner.actor.show();
+        Tweener.addTween(this._spinner.actor,
+                         { opacity: 255,
+                           time: SPINNER_ANIMATION_TIME,
+                           delay: SPINNER_ANIMATION_DELAY,
+                           transition: 'linear' });
+    },
+
+    _stopSpinning: function() {
+        this._spinner.actor.hide();
+        this._spinner.actor.opacity = 0;
+        this._spinner.stop();
+    },
+
+    _reset: function() {
+        this._stopSpinning();
+        this._entry.reset();
+        this._updateSensitivity();
+    },
+
+    _showError: function(error) {
+        if (error.matches(PaygManager.PaygErrorDomain, PaygManager.PaygError.DISABLED))
+            this._setErrorMessage(_("Pay-as-you-go is not enabled in this machine"));
```

Comment:
We probably want to bypass the `PaygUnlockDialog` altogether if something returns the `DISABLED` error, rather than displaying an error message. The idea is that if this error is returned, the user probably isn’t even aware that PAYG is a concept.

Reply:
Well, actually that is the case already (this dialog won't ever show up if PAYG is disabled) so this can't really ever happen. I had it mostly for testing purposes, but I'll remove it.

Reply:
Leave an assertion in its place? Then it’s obvious why the code _doesn’t_ handle `PaygError.DISABLED`.

Reply:
I'm not sure if the equivalent to  `g_assert_not_reached()` is available to GJS, and in any case I feel like we should not crash in that case so I'm adding an `else` branch that would set the `Unknown error` message in the UI, along with a comment in the code to make it explicit that we don't consider any other error.